packages/headless/src/app/ssr-engine/ssr-engine.ts

```diff
@@ -0,0 +1,465 @@
+  [customName: string]: ControllerSnapshot<unknown>;
+}
+
+type AnyAction = {type: string};
+
+export interface EngineSnapshot<
+  TSearchFulfilledAction extends AnyAction,
+  TControllers extends ControllerSnapshotsMap
+> {
+  searchFulfilledAction: TSearchFulfilledAction;
+  controllers: TControllers;
+}
+
+export function mapObject<TKey extends string, TInitialValue, TNewValue>(
+  obj: Record<TKey, TInitialValue>,
+  predicate: (value: TInitialValue, key: TKey) => TNewValue
+): Record<TKey, TNewValue> {
+  return Object.fromEntries(
+    Object.entries(obj).map(([key, value]) => [
+      key,
+      predicate(value as TInitialValue, key as TKey),
+    ])
+  ) as Record<TKey, TNewValue>;
+}
+
+export function defineSearchEngine<
+  TControllerDefinitions extends ControllerDefinitionsMap<
+    SearchEngine,
+    Controller
+  >
+>({
+  controllers: controllerDefinitions,
+  ...engineOptions
+}: SearchEngineDefinitionOptions<TControllerDefinitions>): SearchEngineDefinition<TControllerDefinitions> {
+  const build: SearchEngineDefinition<TControllerDefinitions>['build'] = async (
+    ...[buildOptions]: Parameters<
+      SearchEngineDefinition<TControllerDefinitions>['build']
+    >
+  ) => {
+    const engine = buildSearchEngine(
+      buildOptions?.extend
+        ? await buildOptions.extend(engineOptions)
+        : engineOptions
+    );
+    const controllerOptions =
+      buildOptions && 'controllers' in buildOptions
+        ? buildOptions.controllers
+        : null;
+    const controllers = controllerDefinitions
+      ? mapObject(controllerDefinitions, (definition, key) =>
+          'build' in definition
+            ? definition.build(engine)
+            : definition.buildWithProps(
+                engine,
+                controllerOptions?.[key as keyof typeof controllerOptions]
+              )
+        )
+      : {};
+    return {
+      engine,
+      controllers:
+        controllers as InferControllersMapFromDefinition<TControllerDefinitions>,
+    };
+  };
+
+  const fetchInitialState: SearchEngineDefinition<TControllerDefinitions>['fetchInitialState'] =
+    (
+      ...[executeOptions]: Parameters<
+        SearchEngineDefinition<TControllerDefinitions>['fetchInitialState']
+      >
+    ) =>
+      new Promise<
+        EngineSnapshot<
+          {type: string},
+          InferControllerSnapshotsMapFromDefinitions<TControllerDefinitions>
+        >
+        // eslint-disable-next-line no-async-promise-executor
+      >(async (resolve, reject) => {
+        const middleware: Middleware = () => (next) => (action) => {
+          next(action);
+          if (action.type === 'search/executeSearch/fulfilled') {
+            resolve({
+              controllers: mapObject(controllers, (controller) => ({
+                initialState: controller.state,
+              })) as InferControllerSnapshotsMapFromDefinitions<TControllerDefinitions>,
+              searchFulfilledAction: JSON.parse(JSON.stringify(action)),
```

Comment:
In my local tests, I sent the action because it's easier, but I'm not entirely sure if that's what we should send.
* Pros
   * We don't need to add an extra action and clone the handler of `executeSearch.fulfilled` in every slice.
* Cons
  * To the user, it may look hack-ish? They don't need to know we're sending an action, they just need to serialize it.
  * The user may try to mess around with the action, like replacing parts of the search response.

A possible solution would be to serialize and deserialize the action ourselves. However, if we serialize the action to a string, then Typescript won't stop the implementer from passing some unrelated string to the `hydrateInitialState` function.

Example:
```ts
export type EngineDigest<TEngine extends CoreEngine> = string;

export interface EngineDefinitionHydrateOptionsWithoutProps<TEngine extends CoreEngine> {
  digest: EngineDigest<CoreEngine>;
}
// ...
const {digest} = await engineDefinition.fetchInitialState();
return <MyPage digest={digest} />;
// ...
(props: MyPageProps) => {
  // ...
  const {engine, controllers} = await engineDefinition.hydrateInitialState({ digest: props.digest });
}
```

Reply:
I experimented with using symbols in Typescript to anonymize the type of a property, and it works, but it's finicky...
<img width="566" alt="image" src="https://github.com/coveo/ui-kit/assets/54454747/b1d93a7d-451c-4e79-9eed-c06cc09fa505">

Reply:
Closing this, since I created a separate Jira for it.